src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -321,107 +309,127 @@ private bool TryLookupLengthOrCount(SyntaxNode node, TypeSymbol inputType, out P
+            var bindingDiagnostics = BindingDiagnosticBag.GetInstance(diagnostics);
             TypeSymbol argType = Compilation.GetWellKnownType(argIsIndex ? WellKnownType.System_Index : WellKnownType.System_Range);
-            if (argType.IsErrorType())
-            {
-                // If the argType is missing, we will fallback to the implicit indexer support.
-                return false;
-            }
+            var lookupResult = LookupResult.GetInstance();
+            if (!argType.IsErrorType())
+            {
+                LookupMembersInType(
+                    lookupResult,
+                    receiverType,
+                    WellKnownMemberNames.Indexer,
+                    arity: 0,
+                    basesBeingResolved: null,
+                    LookupOptions.Default,
+                    originalBinder: this,
+                    diagnose: false,
+                    ref useSiteInfo);
+
+                if (lookupResult.IsMultiViable)
+                {
+                    var indexerGroup = ArrayBuilder<PropertySymbol>.GetInstance(lookupResult.Symbols.Count);
+                    foreach (Symbol symbol in lookupResult.Symbols)
+                    {
+                        Debug.Assert(symbol.IsIndexer());
+                        indexerGroup.Add((PropertySymbol)symbol);
+                    }
+                    lookupResult.Clear();
 
-            LookupMembersInType(
-                lookupResult,
-                receiverType,
-                WellKnownMemberNames.Indexer,
-                arity: 0,
-                basesBeingResolved: null,
-                LookupOptions.Default,
-                originalBinder: this,
-                diagnose: false,
-                ref useSiteInfo);
+                    var analyzedArguments = AnalyzedArguments.GetInstance();
+                    analyzedArguments.Arguments.Add(new BoundIndexOrRangeIndexerPatternValuePlaceholder(syntax, argType));
+                    var receiver = new BoundImplicitReceiver(syntax, receiverType);
+                    BoundExpression boundAccess = BindIndexerOrIndexedPropertyAccess(syntax, receiver, indexerGroup, analyzedArguments, bindingDiagnostics);
+                    switch (boundAccess)
+                    {
+                        case BoundIndexerAccess boundIndexerAccess:
+                            if (boundIndexerAccess.ResultKind == LookupResultKind.Viable &&
+                                boundIndexerAccess.Indexer.GetMethod is { } getMethod &&
+                                IsAccessible(getMethod, ref useSiteInfo) &&
+                                TryLookupLengthOrCount(receiverType, lookupResult, out lengthProperty, ref useSiteInfo))
+                            {
+                                // PROTOTYPE(list-patterns) Can this be ever true? If so, move to if above
+                                Debug.Assert(!boundIndexerAccess.Indexer.IsStatic);
```

Comment:
It likely can. We're using `LookupOptions.Default` above, which makes no restrictions on whether the indexer must be static or instance. If we want it to be an instance indexer, we should change the option for that.

Reply:
Default being used elsewhere as well probably because indexers usually can't be static. I couldn't come up with a test case to verify though.

Reply:
IL that looks something like this should do the trick:
```
.class public auto ansi beforefieldinit C
    extends [System.Private.CoreLib]System.Object
{
    .custom instance void [System.Private.CoreLib]System.Reflection.DefaultMemberAttribute::.ctor(string) = (
        01 00 04 49 74 65 6d 00 00
    )
    // Methods
    .method public hidebysig specialname 
        static int32 get_Item (
            int32 i
        ) cil managed 
    {
        // Method begins at RVA 0x2050
        // Code size 2 (0x2)
        .maxstack 8

        IL_0000: ldc.i4.1
        IL_0001: ret
    } // end of method C::get_Item

    .method public hidebysig specialname rtspecialname 
        instance void .ctor () cil managed 
    {
        // Method begins at RVA 0x2053
        // Code size 8 (0x8)
        .maxstack 8

        IL_0000: ldarg.0
        IL_0001: call instance void [System.Private.CoreLib]System.Object::.ctor()
        IL_0006: nop
        IL_0007: ret
    } // end of method C::.ctor

    // Properties
    .property int32 Item(
        int32 i
    )
    {
        .get int32 C::get_Item(int32)
    }

} // end of class C
```

Corresponds roughly to what `static int this[int i]` might look like, if that were a thing that was allowed.